pkg/serviceaccount/openidmetadata.go

```diff
@@ -0,0 +1,283 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package serviceaccount
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"net/url"
+
+	jose "gopkg.in/square/go-jose.v2"
+
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+const (
+	// OpenIDConfigPath is the URL path at which the API server serves
+	// an OIDC Provider Configuration Information document, corresponding
+	// to the Kubernetes Service Account key issuer.
+	// https://openid.net/specs/openid-connect-discovery-1_0.html
+	OpenIDConfigPath = "/.well-known/openid-configuration"
+
+	// JWKSPath is the URL path at which the API server serves a JWKS
+	// containing the public keys that may be used to sign Kubernetes
+	// Service Account keys.
+	JWKSPath = "/openid/v1/jwks"
+)
+
+// OpenIDMetadata contains the pre-rendered responses for OIDC discovery endpoints.
+type OpenIDMetadata struct {
+	ConfigJSON       []byte
+	PublicKeysetJSON []byte
+}
+
+// NewOpenIDMetadata returns the pre-rendered JSON responses for the OIDC discovery
+// endpoints, or an error if they could not be constructed. Callers should note
+// that this function may perform additional validation on inputs that is not
+// backwards-compatible with all command-line validation. The recommendation is
+// to log the error and skip installing the OIDC discovery endpoints.
+func NewOpenIDMetadata(issuerURL, jwksURI, defaultExternalAddress string, pubKeys []interface{}) (*OpenIDMetadata, error) {
+	if issuerURL == "" {
+		return nil, fmt.Errorf("empty issuer URL")
+	}
+	if jwksURI == "" && defaultExternalAddress == "" {
+		return nil, fmt.Errorf("either the JWKS URI or the default external address, or both, must be set")
+	}
+	if len(pubKeys) == 0 {
+		return nil, fmt.Errorf("no keys provided for validating keyset")
+	}
+
+	// Ensure the issuer URL is https, per the OIDC spec (this is the additional
+	// validation the doc comment warns about).
+	iss, err := url.Parse(issuerURL)
+	if err != nil {
+		return nil, err
+	}
+	if iss.Scheme != "https" {
+		return nil, fmt.Errorf("issuer URL must use https scheme, got: %s", issuerURL)
+	}
```

Comment:
also make sure this doesn't have a query or fragment per https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata

> URL using the https scheme with no query or fragment component that the OP asserts as its Issuer Identifier

Reply:
Done and added test cases.